File: apps/backend/src/main/java/timesheets/domain/TimeEntry.java
package timesheets.domain;

import jakarta.persistence.*;
import java.time.LocalDateTime;
import java.util.UUID;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

// so this is the JPA entity- you can kinda think of it like the data template
// it's basically showing us what the time entry looks like in the database
// a java object is what the row looks like in the database

@Entity // this tells Java that the class maps to a DB
@Table(
    name =
        "time_entries") // what the table will be called in the database, otherwise TimeEntry would
// be used
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class TimeEntry {

  @Id
  @GeneratedValue(strategy = GenerationType.UUID)
  private UUID id;

  @Column(name = "timesheet_id")
  private UUID timesheetId;

  @Column(name = "workspace_member_id", nullable = false)
  private UUID workspaceMemberId;

  @Column(name = "project_id", nullable = false)
  private UUID projectId;

  @Column(name = "task_id")
  private UUID taskId;

  @Column(name = "start_time", nullable = false)
  private LocalDateTime startTime;

  @Column(name = "end_time")
  private LocalDateTime endTime;

  @Column(name = "duration_seconds")
  private Integer durationSeconds;

  @Column(name = "entry_type")
  private String entryType;

  private String description;

  @Column(name = "is_deleted")
  @Builder.Default
  private Boolean isDeleted = false;

  @Column(name = "is_locked")
  @Builder.Default
  private Boolean isLocked = false;

  @Column(name = "deleted_at")
  private LocalDateTime deletedAt;

  @Column(name = "created_at", nullable = false, updatable = false)
  private LocalDateTime createdAt;

  @Column(name = "updated_at")
  private LocalDateTime updatedAt;

  @PrePersist // this makes the method run before the entity is saved
  protected void onCreate() {
    createdAt = LocalDateTime.now(); // sets the timestamp before the insert
    updatedAt = LocalDateTime.now();

    if (entryType == null) {
      entryType = "MANUAL";
    }

    if (isDeleted == null) {
      isDeleted = false;
    }
  }

  @PreUpdate // runs before the entity is updated
  protected void onUpdate() {
    updatedAt = LocalDateTime.now();
  }
}

File: ai-service/app/models/time_entry.py
"""
This is the SQLAlchemy model for time entries. its built from the real schema in backend
namely V1

Author: Zamokuhle Zwane
Date: 21/08/2026
"""

import uuid
from datetime import datetime

from sqlalchemy import Boolean, DateTime, ForeignKey, Integer, String, Text
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.database import Base


class TimeEntry(Base):
    __tablename__ = "time_entries"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    timesheetid: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("timesheets.id"), nullable=True
    )
    workspace_member_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("workspace_members.id")
    )
    project_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("projects.id"))
    task_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("tasks.id"), nullable=True
    )
    start_time: Mapped[datetime] = mapped_column(DateTime)
    end_time: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    duration_seconds: Mapped[int | None] = mapped_column(Integer, nullable=True)
    entry_type: Mapped[str] = mapped_column(String(10), default="MANUAL")
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    is_locked: Mapped[bool] = mapped_column(Boolean, default=False)
    is_deleted: Mapped[bool] = mapped_column(Boolean, default=False)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)
